### PR TITLE
Modified Dockerfile for multi stage build and GHA to add the build arg

### PR DESCRIPTION
**Pull request by DipayanP007**
Solves the issue 
#35 

* I created a Dockerfile for multi stage build
* Added build args to pass `Github SHA` and `openweather API key` to the Dockerfile
*  Changed the line in GHA to use build args for `Github SHA` and `openweather API`

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -1,4 +1,28 @@
+FROM mcr.microsoft.com/dotnet/sdk:3.1 as builder
+
+ARG GITHUBSHA OPENWEATHERMAP_APIKEY_RAW
```
Comment (torbencarstens) [inline Dockerfile]: This won't work, `ARG` only takes a single argument (see https://docs.docker.com/engine/reference/builder/#arg). This needs to be split into separate lines

Additionally, `OPENWEATHER_APIKEY_RAW` should use an empty value `""` as a default. The api key is not needed for this application to work, I'll make sure (in another PR) that this is set in the github pipeline.

File: Dockerfile [hunk 2]
```diff
@@ -1,4 +1,28 @@
+FROM mcr.microsoft.com/dotnet/sdk:3.1 as builder
+
+ARG GITHUBSHA OPENWEATHERMAP_APIKEY_RAW 
+
+WORKDIR /prelim-frontend
+
+ENV NODE_OPTIONS="--openssl-legacy-provider"
+
+RUN apt update -y && curl -sL https://deb.nodesource.com/setup_17.x | bash -
```
Comment (torbencarstens) [inline Dockerfile]: Please use apt-get here as well.

File: Dockerfile [hunk 3]
```diff
@@ -1,4 +1,28 @@
...
+
+RUN sed -i -e "s#{{TAG}}#$GITHUBSHA#g" src/App.fs && sed -i -e "s/|> Program.withConsoleTrace//g" src/App.fs && sed -i -e "s#{{API_KEY}}#$OPENWEATHERMAP_APIKEY_RAW#g" src/App.fs
+
+RUN npm install && npm run prod
+
+
+# Second stage
 FROM nginx:1.21-alpine
+
+WORKDIR /preliminary-frontend
```
Comment (torbencarstens) [inline Dockerfile]: We don't need a `WORKDIR` here

File: Dockerfile [hunk 4]
```diff
@@ -1,4 +1,28 @@
+FROM mcr.microsoft.com/dotnet/sdk:3.1 as builder
+
+ARG GITHUBSHA OPENWEATHERMAP_APIKEY_RAW 
+
+WORKDIR /prelim-frontend
```
Comment (torbencarstens) [inline Dockerfile]: Please use `/app` here, it's more or less the default for `WORKDIR` by now

File: Dockerfile [hunk 5]
```diff
@@ -1,4 +1,28 @@
...
+
+ENV NODE_OPTIONS="--openssl-legacy-provider"
+
+RUN apt update -y && curl -sL https://deb.nodesource.com/setup_17.x | bash -
+
+RUN apt-get install -y nodejs
+
+COPY . .
+
+RUN sed -i -e "s#{{TAG}}#$GITHUBSHA#g" src/App.fs && sed -i -e "s/|> Program.withConsoleTrace//g" src/App.fs && sed -i -e "s#{{API_KEY}}#$OPENWEATHERMAP_APIKEY_RAW#g" src/App.fs
```
Comment (torbencarstens) [inline Dockerfile]: This should be split into multiple lines (keep it in a single `RUN` statement) for improved readability.

File: Dockerfile [hunk 6]
```diff
@@ -1,4 +1,28 @@
...
+
+RUN npm install && npm run prod
+
+
+# Second stage
 FROM nginx:1.21-alpine
+
+WORKDIR /preliminary-frontend
+
+COPY --from=builder /prelim-frontend/public/ /usr/share/nginx/html
```
Comment (torbencarstens) [inline Dockerfile]: Make sure to use `/app` here as well after changing it in the builder container

File: Dockerfile [hunk 7]
```diff
@@ -1,4 +1,28 @@
+FROM mcr.microsoft.com/dotnet/sdk:3.1 as builder
+
+ARG GITHUBSHA OPENWEATHERMAP_APIKEY_RAW 
+
+WORKDIR /prelim-frontend
+
+ENV NODE_OPTIONS="--openssl-legacy-provider"
```
Comment (torbencarstens) [inline Dockerfile:12]: Please add a comment (with a link to the node17 release docs) that this is due to the openssl 3.0 change. This way it's easier to check in the future whether we can remove this arg or not.

File: Dockerfile [hunk 8]
```diff
@@ -1,4 +1,28 @@
...
+RUN apt-get install -y nodejs
+
+COPY . .
+
+RUN sed -i -e "s#{{TAG}}#$GITHUBSHA#g" src/App.fs && sed -i -e "s/|> Program.withConsoleTrace//g" src/App.fs && sed -i -e "s#{{API_KEY}}#$OPENWEATHERMAP_APIKEY_RAW#g" src/App.fs
+
+RUN npm install && npm run prod
+
+
+# Second stage
```
Comment (torbencarstens) [inline Dockerfile]: You can remove this comment, this is obvious by the `FROM` in the next line.

**Review by torbencarstens — CHANGES_REQUESTED** [review]
@DipayanP007 thanks for your contribution. Please have a look at the comments and feel free to ask if you've any questions about them.

In my opinion this can be merged after the requested changes have been done (as long as it builds correctly).

**Comment by DipayanP007** [comment]
Hey @torbencarstens I have made the requested changes.

**Review by torbencarstens — APPROVED** [review]

**Comment by torbencarstens** [comment]
Thanks for contributing @DipayanP007

**Comment by torbencarstens** [comment]
Closes #35

**Pull request merged**